src/gtc/cuir/extent_analysis.py

```diff
@@ -0,0 +1,120 @@
+# -*- coding: utf-8 -*-
+#
+# GTC Toolchain - GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from collections import defaultdict
+from typing import Any, Dict
+
+from eve import NodeTranslator
+
+from . import cuir
+
+
+class ComputeExtents(NodeTranslator):
+    def visit_VerticalLoopSection(
+        self, node: cuir.VerticalLoopSection, **kwargs: Any
+    ) -> cuir.VerticalLoopSection:
+        horizontal_executions = []
+        extents_map: Dict[str, cuir.IJExtent] = dict()
+        for horizontal_execution in reversed(node.horizontal_executions):
+            writes = (
+                node.iter_tree()
+                .if_isinstance(cuir.AssignStmt)
+                .getattr("left")
+                .if_isinstance(cuir.FieldAccess, cuir.IJCacheAccess)
+                .getattr("name")
+                .to_set()
+            )
+            extent: cuir.IJExtent = cuir.IJExtent.zero().union(
+                *(extents_map.get(write, cuir.IJExtent.zero()) for write in writes),
+            )
+
+            horizontal_executions.append(
+                cuir.HorizontalExecution(
+                    body=horizontal_execution.body,
+                    mask=horizontal_execution.mask,
+                    declarations=horizontal_execution.declarations,
+                    extent=extent,
+                )
+            )
+
+            accesses_map: Dict[str, cuir.IJExtent] = (
+                horizontal_execution.iter_tree()
+                .if_isinstance(cuir.FieldAccess, cuir.IJCacheAccess)
+                .reduceby(
+                    lambda acc, x: acc.union(extent + cuir.IJExtent.from_offset(x.offset)),
+                    "name",
+                    init=cuir.IJExtent.zero(),
+                    as_dict=True,
+                )
+            )
+            extents_map = {
+                k: extents_map.get(k, cuir.IJExtent.zero()).union(
+                    accesses_map.get(k, cuir.IJExtent.zero())
+                )
+                for k in set(extents_map.keys()) | set(accesses_map.keys())
+            }
+
+        return cuir.VerticalLoopSection(
+            start=node.start,
+            end=node.end,
+            horizontal_executions=list(reversed(horizontal_executions)),
+        )
+
+
+class CacheExtents(NodeTranslator):
```

Comment:
Same question here? Should this analysis be done on the oir level? Currently, the only backend consuming this information would be this one, but still feels like it should be done at the higher level.

Reply:
See above.